### PR TITLE
Fix transformation synchronization and HDF5 y-transform (#509, #1003) and update UNet references (#1031)

### LINKED ISSUE #509: Memory Issue with denoisers [Possibly an issue with HDF5 dataset]
HDF5 datasets seem to be causing memory issues in deepinverse.

Issue with following code :

```
from huggingface_hub import hf_hub_download
path = hf_hub_download(repo_id="jtachella/finetuning", repo_type="dataset", filename="cryo_em/dinv_dataset0.h5")
patch_size = 128
train_dataset = dinv.datasets.HDF5Dataset(path=path, transform=transforms.RandomCrop(patch_size))
test_dataset = dinv.datasets.HDF5Dataset(path=path)
train_dataloader = DataLoader(train_dataset, batch_size = 32, shuffle = True)#transforms= transforms.RandomCrop(patch_size))
```

Custom dataset that works :

```
for a,c in enumerate(train_dataloader):
    next
c[0].shape

from torch.utils.data import Dataset
class RandomCropPatchDataset(Dataset):
    def __init__(self, images, patch_size=128, transform=None):
        """
        Args:
            images (list of PIL.Image or torch.Tensor): List of 5 images.
            patch_size (int): Size of the square crop.
            transform (callable, optional): Optional transform to be applied on a sample.
        """
        self.images = images
        self.patch_size = patch_size
        self.transform = transform or transforms.Compose([
            transforms.RandomCrop(patch_size)
            #transforms.ToTensor()
        ])

    def __len__(self):
        return len(self.images) * 100  # You can choose how many patches you want overall

    def __getitem__(self, idx):
        img = self.images[idx % len(self.images)]  # loop over images
        return self.transform(img), self.transform(img)

my_data = RandomCropPatchDataset(images=c[0],patch_size=128)
my_dataloader = DataLoader(my_data, batch_size = 32, shuffle = True)
```

Trainer code : 
```
model = dinv.models.ArtifactRemoval(dinv.models.DRUNet(in_channels=1,out_channels=1,device=device))
loss = dinv.loss.Neighbor2Neighbor()

# choose optimizer and scheduler
learning_rate = 5e-4
epochs = 20
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=1e-8)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=int(epochs * 0.8) + 1)
physics = dinv.physics.Denoising()
trainer = dinv.Trainer(
    model=model,
    physics=physics,
    epochs=epochs,
    scheduler=scheduler,
    losses=loss,
    optimizer=optimizer,
    device=device,
    train_dataloader=my_dataloader,        # train_dataloader -> gives a memory error
    #eval_dataloader=test_dataloader,
    plot_images=True,
    save_path=f"/home/abhishek/Dinv_workshop/ckpts/",
    verbose=True,
    show_progress_bar=True,  # disable progress bar for better vis in sphinx gallery.
    #wandb_vis=wandb_vis,
)
model = trainer.train()

```

Can you please take a look at it ?

Thanks
Abhishek

### LINKED ISSUE #1003: Dataset transforms x not same as y
### Describe the bug

When using deepinv ImageFolder dataset that returns x & y images, if passing a random transform, x and y draw separate instances of the random transform and therefore are transformed differently. This means that may no longer be paired.

- In general, it's non trivial how a transform defined on x is defined on y - should we even allow transforms on paired data?
- Note that using a deepinv transform we can perform reproducible transforms by fixing `params` of the transform. However when using torchvision-style data augmentations, this is impossible.

### Steps to reproduce

let's assume x and y are the same image:
```
import torch
import deepinv as dinv
x = y = torch.randn((1, 1, 8, 8))
torch.save(x, "temp_x.pt")
torch.save(y, "temp_y.pt")
```

Then the following fails, if a random transform is passed in the `transform` parameter:
```
dataset = dinv.datasets.ImageFolder(".", x_path="*_x.pt", y_path="*_y.pt", loader=torch.load, transform=dinv.transform.Rotate())
x, y = dataset[0]
assert torch.all(x == y)
```

### Acceptance criteria

- ImageFolder allows transform that transforms x and y in the same way
- HDF5Dataset transforms x and y in the same way

### LINKED ISSUE #1031: Reference the original design of the UNet
### Documentation location

UNet

### Describe the issue or suggestion

Our UNet is pretty much exactly the one described by Jin et al. (2017) but we don't reference it. I suggest we do

For reference, it differs from the one from Ronneberger et al. (2015) in multiple ways:
- Use of zero padding instead of valid padding
- Introduction of a global residual connection
- Suggested combination with a backprojection / pseudo-inverse
- Target inverse problems specifically instead of medical segmentation

Jin, K. H., McCann, M. T., Froustey, E., & Unser, M. (2017). Deep convolutional neural network for inverse problems in imaging. IEEE transactions on image processing, 26(9), 4509-4522.

Ronneberger, O., Fischer, P., & Brox, T. (2015, October). U-net: Convolutional networks for biomedical image segmentation. In International Conference on Medical image computing and computer-assisted intervention (pp. 234-241). Cham: Springer international publishing.

### Suggested fix or improvement (optional)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rusheel86**
This PR addresses two linked issues regarding dataset transformations:
Issue #509: Added the missing self.transform(y) call in HDF5Dataset.__getitem__ to ensure measurements are transformed alongside ground truth.
Issue #1003: Implemented RNG state synchronization in both ImageFolder and HDF5Dataset. This ensures that random augmentations (e.g., RandomRotation) are applied identically to both $x$ and $y$.
Issue #1031: Updated UNet docstring to correctly reference Jin et al. (2017) and added the corresponding BibTeX entry to refs.bib.

Changes:
Modified deepinv/datasets/base.py: Captured and reset torch.get_rng_state() in ImageFolder.__getitem__.
Modified deepinv/datasets/datagenerator.py: Implemented y transformation and RNG synchronization in HDF5Dataset.__getitem__.
Modified deepinv/models/unet.py: Updated docstring for scientific accuracy.
Modified docs/source/refs.bib: Added Jin et al. (2017) citation.

Testing:
Verified locally with a synchronization script ensuring torch.allclose(x, y) after random rotation.

### CODE REVIEW

**Comment by jscanvic** [comment]
Welcome in and thanks for taking care of those issues!

We probably better merge #764 before this PR - I'll make the final changes asap

**Comment by Rusheel86** [comment]
Happy to contribute! That sounds good so I’ll wait for #764 to be merged and once that's in, I'll rebase my changes on top of the revamp and resolve the conflicts. In the meantime, I've just fixed the duplicate bibliography entry and confirmed the linting passes so it passes all checks.

**Comment by Andrewwango** [comment]
Thanks for the PR @Rusheel86 ! I will review it as soon as it's sync'd with #764 , no action on your side yet :)

**Comment by Andrewwango** [comment]
Hi @Rusheel86 , please can you rebase now that #764 is merged, fix the conflicts, and we can go from there?
Thanks again for contributing!

**Comment by Rusheel86** [comment]
Hello @Andrewwango, I'm glad to contribute. Rebased and conflicts fixed so let me know if everything looks good on your end now. Thank you.
I'm still relatively new to open source so I'm sticking to smaller fixes for now and looking for more ways to contribute. Do you have any specific issues or areas you’d recommend I look into next?

**Comment by Rusheel86** [comment]
@Andrewwango sorry for the multiple commits, all checks are passing except the lint job, which is failing due to some GitHub API issue (uv install 404 error, unrelated to my changes). Could you re-run it when you get a chance?

**Comment by Rusheel86** [comment]
Hi @Andrewwango, all CI checks are now passing, please let me know if there are any other adjustments needed before merging.

**Comment by Rusheel86** [comment]
Hello @Andrewwango, sorry for the ping,  please let me know if the pr needs anything else.
Thank you

File: deepinv/models/unet.py [hunk 1]
```diff
@@ -14,14 +14,13 @@ class UNet(Denoiser):
...
     U-Net convolutional denoiser.
 
-    This network is a fully convolutional denoiser based on the U-Net architecture. The number of stages in the network is
-    controlled by ``scales``. The width of each stage is controlled by ``channels_per_scale``,
-    which gives the number of feature maps at each stage, from shallow to deeper stages.
-    The number of trainable parameters increases with both ``scales`` and the values in ``channels_per_scale``.
-
-    If ``scales`` is not given, it is inferred from ``channels_per_scale``. If both are omitted, defaults to
-    ``channels_per_scale=[64, 128, 256, 512]``. If only ``scales`` is specified, ``channels_per_scale=[64 * (2**k) for k in range(scales)]``.
-    When both are specified, ``scales`` must match the length of ``channels_per_scale``.
```
Comment (Andrewwango) [inline deepinv/models/unet.py:24]: why have these lines been deleted?

File: deepinv/datasets/base.py [hunk 2]
```diff
@@ -376,24 +376,62 @@ def __init__(
...
+        """
+        Returns the sample at the given index.
+
+        :param int idx: index of the sample to load.
+        :return: transformed x, y, and optionally params.
+        """
+
+        # Load x and y
+        x = (
+            torch.tensor(float("nan"))
```
Comment (Andrewwango) [inline deepinv/datasets/base.py]: torch.nan

File: deepinv/datasets/base.py [hunk 3]
```diff
@@ -376,24 +376,62 @@ def __init__(
...
+            if self.x_paths is None
+            else self.loader(self.x_paths[idx])
+        )
+        y = None if self.y_paths is None else self.loader(self.y_paths[idx])
 
-        if self.y_paths is None:
-            y = None
+        is_supervised = not getattr(self, "unsupervised", False)
```
Comment (Andrewwango) [inline deepinv/datasets/base.py]: `unsupervised` is never set - what are you doing here?

Please rewrite this section manually and carefully, since it seems like this variable has been hallucinated.

File: deepinv/datasets/base.py [hunk 4]
```diff
@@ -376,24 +376,62 @@ def __init__(
...
+                y = self.transform_y(y)
+
+        # Ensure x is a tensor
+        if self.estimate_params is not None:
+            x_for_params = x
+            if not isinstance(x_for_params, torch.Tensor):
+                from torchvision.transforms import ToTensor
+
+                x_for_params = ToTensor()(x_for_params)
+            params = self.estimate_params(x_for_params, y)
```
Comment (Andrewwango) [inline deepinv/datasets/base.py]: why are you adding this?

this was not present in the original code, and this is not solving any issues stated in the PR.

the ToTensor usage is inappropriate.

File: deepinv/datasets/datagenerator.py [hunk 5]
```diff
@@ -367,48 +368,54 @@ def update_split_size(size: int, *, field_name: str) -> None:
             deprecation_message="The attribute 'data_cache' is deprecated and will be removed in future versions.",
         )
 
-    def __getitem__(self, index: int) -> tuple:
```
Comment (Andrewwango) [inline deepinv/datasets/datagenerator.py:329]: There are many diffs here which feel unnecessary. Please carefully review your changes to make sure you're not breaking existing code.

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thanks for your PR @Rusheel86  and apologies for the delayed review. Please address the comments in the review, and let me know when the PR is ready again.

**Comment by Rusheel86** [comment]
Sorry about that, will convert back once I fix everything for this pr.

**Comment by Rusheel86** [comment]
Thanks for the review even though the pr was pretty messy as I changed a lot of things trying to pass test_datasets.py locally. I've made the suggested changes as best as I could.

**Comment by Rusheel86** [comment]
@Andrewwango thanks for the feedback, the pr is ready to review whenever you have time.

File: deepinv/models/unet.py [hunk 6]
```diff
@@ -14,11 +14,15 @@ class UNet(Denoiser):
     r"""
     U-Net convolutional denoiser.
 
-    This network is a fully convolutional denoiser based on the U-Net architecture. The number of stages in the network is
-    controlled by ``scales``. The width of each stage is controlled by ``channels_per_scale``,
-    which gives the number of feature maps at each stage, from shallow to deeper stages.
-    The number of trainable parameters increases with both ``scales`` and the values in ``channels_per_scale``.
+    The architecture follows the design described in :footcite:t:`jin2017deep`, which is adapted for
```
Comment (Andrewwango) [inline deepinv/models/unet.py:17]:
```suggestion
    This network is a fully convolutional denoiser based on the U-Net architecture. The architecture follows the design described in :footcite:t:`jin2017deep`, which is adapted for
```

File: deepinv/datasets/base.py [hunk 7]
```diff
@@ -377,18 +377,21 @@ def __len__(self):
...
-            y = self.transform_y(self.loader(self.y_paths[idx]))
+        # Capture RNG state to synchronize paired random transforms
+        state = torch.get_rng_state()
 
-        params = self.estimate_params(x, y) if self.estimate_params is not None else {}
+        if self.x_paths is not None and self.transform_x is not None:
+            x = self.transform_x(x)
+
+        if y is not None and self.transform_y is not None:
+            torch.set_rng_state(state)
```
Comment (Andrewwango) [inline deepinv/datasets/base.py:391]: Thanks, I like this solution. Can you add a unit test that this solves the issue? i.e. using the reproducing steps in https://github.com/deepinv/deepinv/issues/1003

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thank you @Rusheel86  and sorry for the delay in reviewing! This PR is almost there. Please can you:

- address the minor changes requested below
- fix the merge conflicts.
- add the transform rng fix to changelog, since it's quite important

Then I can approve and we'll be good to go! Thanks!

**Comment by Rusheel86** [comment]
@Andrewwango Thank you, please a look and let me know if anything else is left

**Review by Andrewwango — APPROVED** [review]
LGTM, thanks!
